src/charm.py

```diff
@@ -1,427 +1,269 @@
-            raise CheckFailedError(f"{e.status.message}", e.status_type)
-        return image_details
-
-    def _validate_default_s3_bucket(self, obj_storage):
-        """Validates the default S3 store, ensuring bucket is accessible and creating if needed."""
-        # Validate the bucket name
-        bucket_name = self.config["default_artifact_root"]
-        if not validate_s3_bucket_name(bucket_name):
-            msg = (
-                f"Invalid value for config default_artifact_root '{bucket_name}'"
-                f" - value must be a valid S3 bucket name"
-            )
-            raise CheckFailedError(msg, BlockedStatus)
-
-        # Ensure the bucket exists, creating it if missing and create_root_if_not_exists==True
-        s3_wrapper = S3BucketWrapper(
-            access_key=obj_storage["access-key"],
-            secret_access_key=obj_storage["secret-key"],
-            s3_service=f"{obj_storage['service']}.{obj_storage['namespace']}",
-            s3_port=obj_storage["port"],
-        )
-
-        if s3_wrapper.check_if_bucket_accessible(bucket_name):
-            return bucket_name
-        else:
-            if self.config["create_default_artifact_root_if_missing"]:
-                try:
-                    s3_wrapper.create_bucket(bucket_name)
-                    return bucket_name
-                except Exception as e:
-                    raise CheckFailedError(
-                        "Error with default S3 artifact store - bucket not accessible or "
-                        f"cannot be created.  Caught error: '{str(e)}",
-                        BlockedStatus,
-                    )
-            else:
-                raise CheckFailedError(
-                    "Error with default S3 artifact store - bucket not accessible or does not "
-                    "exist. Set create_default_artifact_root_if_missing=True to automatically "
-                    "create a missing default bucket",
-                    BlockedStatus,
-                )
-
-    def _define_secrets(self, obj_storage, mysql):
-        """Returns needed secrets in pod_spec.kubernetesResources.secrets format."""
-        return [
-            {
-                "name": f"{self.charm_name}-minio-secret",
-                "data": _minio_credentials_dict(obj_storage=obj_storage),
-            },
-            {
-                "name": f"{self.charm_name}-seldon-init-container-s3-credentials",
-                "data": _seldon_credentials_dict(obj_storage=obj_storage),
-            },
-            {"name": f"{self.charm_name}-db-secret", "data": _db_secret_dict(mysql=mysql)},
-        ]
+            relational_db = relational_db[0]
+            db_unit = list(relational_db.units)[0]
+            relational_db = relational_db.data[db_unit]
+            relational_db["database"]
+            return relational_db
+        except (IndexError, KeyError):
+            raise ErrorWithStatus("Waiting for relational-db relation data", WaitingStatus)
```

Comment:
What about splitting these up to be more descriptive?  I think this catches both:
* we have no related charm on `relational-db`
* we have a related charm, but it hasn't sent us the data we expected

Can we parse these separately to be clearer about this?  It feels like it would be hard to debug.  I think `IndexError` will always mean we have no relation, and `KeyError` that we have a relation but it hasn't sent us what we expect?